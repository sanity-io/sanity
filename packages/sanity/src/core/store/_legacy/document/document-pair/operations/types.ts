import {SanityDocument, Schema} from '@sanity/types'
import {SanityClient} from '@sanity/client'
import {IdPair} from '../../types'
import {DocumentVersionSnapshots} from '../snapshotPair'
import {HistoryStore} from '../../../history'

/** @internal */
export interface OperationImpl<
  ExtraArgs extends any[] = [],
  DisabledReason extends string = string,
> {
  disabled: (args: OperationArgs) => DisabledReason | 'NOT_READY' | false
  execute(args: OperationArgs, ...extra: ExtraArgs): void
}

/** @internal */
export interface Operation<ExtraArgs extends any[] = [], ErrorStrings extends string = string> {
  disabled: false | ErrorStrings | 'NOT_READY'
  execute(...extra: ExtraArgs): void
}

type GuardedOperation = Operation<any[], 'NOT_READY'>
type Patch = any

/** @internal */
// Note: Changing this interface in a backwards incompatible manner will be a breaking change
export interface OperationsAPI {
  commit: Operation | GuardedOperation
  delete: Operation<[], 'NOTHING_TO_DELETE' | 'NOT_READY'>
  del: Operation<[], 'NOTHING_TO_DELETE'> | GuardedOperation
  publish:
    | Operation<[], 'LIVE_EDIT_ENABLED' | 'ALREADY_PUBLISHED' | 'NO_CHANGES'>
    | GuardedOperation
  patch: Operation<[patches: Patch[], initialDocument?: Record<string, any>]> | GuardedOperation
  discardChanges: Operation<[], 'NO_CHANGES' | 'NOT_PUBLISHED'> | GuardedOperation
  unpublish: Operation<[], 'LIVE_EDIT_ENABLED' | 'NOT_PUBLISHED'> | GuardedOperation
  duplicate: Operation<[documentId: string], 'NOTHING_TO_DUPLICATE'> | GuardedOperation
  restore: Operation<[revision: string]> | GuardedOperation
}

/** @internal */
export interface OperationArgs {
  historyStore: HistoryStore
  client: SanityClient
  schema: Schema
  typeName: string
  idPair: IdPair
  snapshots: {draft: null | SanityDocument; published: null | SanityDocument}
  draft: DocumentVersionSnapshots
  published: DocumentVersionSnapshots
}
