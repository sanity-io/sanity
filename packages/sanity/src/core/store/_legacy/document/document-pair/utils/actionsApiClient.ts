import {type SanityClient} from 'sanity'

const ACTIONS_API_MINIMUM_VERSION = '2024-05-23'

export function actionsApiClient(client: SanityClient): SanityClient {
  return client.withConfig({
    apiVersion: ACTIONS_API_MINIMUM_VERSION,
  })
}
