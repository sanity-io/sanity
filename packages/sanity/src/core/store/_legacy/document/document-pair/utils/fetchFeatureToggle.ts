import {type SanityClient} from '@sanity/client'
import {map, type Observable, of, ReplaySubject, timeout, timer} from 'rxjs'
import {catchError, concatMap, share} from 'rxjs/operators'

interface ActionsFeatureToggle {
  actions: boolean
}
const CACHE = new WeakMap<SanityClient, Observable<boolean>>()

// How often to refresh the feature toggle
const REFRESH_INTERVAL = 1000 * 120

// Timer used to reset the observable when it completes or it's refcount drops to zero
const RESET_TIMER = timer(REFRESH_INTERVAL)

function createFeatureToggle(client: SanityClient) {
  const {dataset} = client.config()

  return timer(0, REFRESH_INTERVAL).pipe(
    concatMap(() =>
      client.observable
        .request({
          uri: `/data/actions/${dataset}`,
          withCredentials: true,
        })
        .pipe(
          map((res: ActionsFeatureToggle) => res.actions),
          timeout({first: 2000, with: () => of(false)}),
          catchError(() =>
            // If we fail to fetch the feature toggle, we'll just assume it's disabled and fallback to legacy mutations
            of(false),
          ),
        ),
    ),
    share({
      // replay latest known state to new subscribers
      connector: () => new ReplaySubject(1),
      // this will typically be completed and unsubscribed from right after the answer is received, so we don't want to reset
      resetOnComplete: () => RESET_TIMER,
      // keep it alive for some time after the last subscriber unsubscribes
      resetOnRefCountZero: () => RESET_TIMER,
    }),
  )
}

export const fetchFeatureToggle = (client: SanityClient): Observable<boolean> => {
  if (!CACHE.has(client)) {
    CACHE.set(client, createFeatureToggle(client))
  }
  return CACHE.get(client)!
}
