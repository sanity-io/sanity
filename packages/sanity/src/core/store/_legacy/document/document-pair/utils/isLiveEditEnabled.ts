import {type Schema} from '@sanity/types'

export const isLiveEditEnabled = (schema: Schema, typeName: string): boolean =>
  (globalThis as any).__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED_LIVE_EDIT_OVERRIDE ??
  schema.get(typeName)?.liveEdit === true
