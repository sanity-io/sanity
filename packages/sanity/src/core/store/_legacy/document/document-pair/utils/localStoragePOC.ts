import {isSanityDocument, type SanityDocument} from '@sanity/types'

import {supportsLocalStorage} from '../../../../../util/supportsLocalStorage'
import {type IdPair} from '../../types'

const createDocumentLocalStorageKey = (documentId: string) => `sanity:editState:${documentId}`

const getDocumentFromLocalStorage = (id: string): SanityDocument | null => {
  if (!supportsLocalStorage) return null

  const key = createDocumentLocalStorageKey(id)

  try {
    const document = localStorage.getItem(key)

    if (!document) return null
    const parsed = JSON.parse(document)
    return isSanityDocument(parsed) ? parsed : null
  } catch (error) {
    console.error(`Error parsing document with ID ${id} from localStorage:`, error)
    return null
  }
}

const saveDocumentToLocalStorage = (document: SanityDocument) => {
  if (!supportsLocalStorage) return

  const key = createDocumentLocalStorageKey(document._id)

  try {
    localStorage.setItem(key, JSON.stringify(document))
  } catch (error) {
    console.error(`Error saving document with ID ${document._id} to localStorage:`, error)
  }
}

/**
 * Function to get the draft and published document from local storage
 * it's not production ready, it's POC only, local storage supports up to 5mb of data which won't be enough for the datasets.
 * @internal
 * @hidden
 */
export const getPairFromLocalStorage = (idPair: IdPair) => {
  if (!supportsLocalStorage) {
    return {
      draft: null,
      published: null,
    }
  }

  return {
    draft: getDocumentFromLocalStorage(idPair.draftId),
    published: getDocumentFromLocalStorage(idPair.publishedId),
  }
}

/**
 * Function to save the draft and published documents to local storage.
 * Note: This is a proof of concept and not production-ready.
 * Local storage supports up to 5MB of data, which will not be sufficient for large datasets.
 * @internal
 * @hidden
 */
export const savePairToLocalStorage = (
  documentPair: {
    draft: SanityDocument | null
    published: SanityDocument | null
  } | null,
) => {
  if (documentPair?.draft) {
    saveDocumentToLocalStorage(documentPair.draft)
  }
  if (documentPair?.published) {
    saveDocumentToLocalStorage(documentPair.published)
  }
}
