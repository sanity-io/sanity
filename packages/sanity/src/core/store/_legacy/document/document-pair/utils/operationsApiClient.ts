import {type SanityClient} from '@sanity/client'

import {RELEASES_STUDIO_CLIENT_OPTIONS} from '../../../../../releases/util/releasesClient'
import {type IdPair} from '../../types'

export function operationsApiClient(client: SanityClient, idPair: IdPair): SanityClient {
  if (idPair.versionId) {
    // TODO: Remove after API version is stable and support releases
    return client.withConfig(RELEASES_STUDIO_CLIENT_OPTIONS)
  }
  return client
}
