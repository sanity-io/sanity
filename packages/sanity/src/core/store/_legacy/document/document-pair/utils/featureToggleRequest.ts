//this is an example function -- we're just calling a document

import {type SanityClient} from '@sanity/client'
import {map, type Observable, of} from 'rxjs'

import {type IdPair} from '../../types'
import {memoize} from '../../utils/createMemoizer'

//in the "real" code, this would be observable.request, to a URI
export const featureToggleRequest: (
  client: SanityClient,
  idPair: IdPair,
  serverActionsEnabled: boolean,
) => Observable<boolean> = memoize(
  (client: SanityClient, idPair: IdPair, serverActionsEnabled: boolean): Observable<boolean> => {
    if (!serverActionsEnabled) {
      return of(false)
    }
    return client.observable
      .fetch('*[_id == $id][0]', {id: '20449512-1e9c-44f0-a509-417c901fbbbd'})
      .pipe(map((doc) => doc.name.split('CONTROL FOR SERVER ACTIONS: ')[1] === 'enabled'))
  },
  (client: SanityClient, idPair: IdPair, serverActionsEnabled: boolean) => {
    const config = client.config()
    return `${config.dataset ?? ''}-${config.projectId ?? ''}-${idPair.publishedId}-${serverActionsEnabled ? '-serverActionsEnabled' : ''}`
  },
)
