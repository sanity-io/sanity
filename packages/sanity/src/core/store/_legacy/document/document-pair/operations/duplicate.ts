import {type SanityDocument, type SanityDocumentLike} from '@sanity/types'
import {omit} from 'lodash'

import {getDraftId, getVersionFromId, getVersionId} from '../../../../../util'
import {isLiveEditEnabled} from '../utils/isLiveEditEnabled'
import {operationsApiClient} from '../utils/operationsApiClient'
import {type MapDocument, type OperationImpl} from './types'

const omitProps = ['_createdAt', '_updatedAt']

const getDocumentToDuplicateId = ({
  versionSnapshot,
  dupeId,
  liveEdit,
}: {
  versionSnapshot?: SanityDocument | null | undefined
  dupeId: string
  liveEdit: boolean
}) => {
  if (versionSnapshot) {
    // When duplicating a version document we need to create it with a version id.
    // We get the version from the snapshot id and create a new version id for the duplicate.
    const versionId = getVersionFromId(versionSnapshot._id)
    if (versionId) return getVersionId(dupeId, versionId)
  }

  if (liveEdit) {
    return dupeId
  }

  return getDraftId(dupeId)
}

const mapDocumentNoop: MapDocument = (document) => document

export const duplicate: OperationImpl<
  [
    baseDocumentId: string,
    options?: {
      mapDocument?: MapDocument
    },
  ],
  'NOTHING_TO_DUPLICATE'
> = {
  disabled: ({snapshots}) => {
    return snapshots.published || snapshots.draft || snapshots.version
      ? false
      : 'NOTHING_TO_DUPLICATE'
  },
  execute: (
    {schema, client, snapshots, typeName, idPair},
    dupeId,
    {mapDocument = mapDocumentNoop} = {},
  ) => {
    const source = snapshots.version || snapshots.draft || snapshots.published

    if (!source) {
      throw new Error('cannot execute on empty document')
    }

    const _id = getDocumentToDuplicateId({
      versionSnapshot: snapshots.version,
      dupeId,
      liveEdit: isLiveEditEnabled(schema, typeName),
    })

    return operationsApiClient(client, idPair).observable.create(
      omit(
        mapDocument({
          ...source,
          _id,
          _type: source._type,
        }),
        omitProps,
        // Performing `Omit` on a type that has an index signature causes the known properties to
        // be lost.
      ) as SanityDocumentLike,
      {
        tag: 'document.duplicate',
      },
    )
  },
}
