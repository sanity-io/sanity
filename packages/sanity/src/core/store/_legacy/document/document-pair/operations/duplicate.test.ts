import {type SanityDocument} from '@sanity/types'
import {beforeEach, describe, expect, it, type Mock, vi} from 'vitest'

import {createMockSanityClient} from '../../../../../../../test/mocks/mockSanityClient'
import {isLiveEditEnabled} from '../utils/isLiveEditEnabled'
import {duplicate} from './duplicate'
import {type OperationArgs} from './types'

vi.mock('../utils/isLiveEditEnabled', () => ({isLiveEditEnabled: vi.fn()}))

beforeEach(() => {
  ;(isLiveEditEnabled as Mock).mockClear()
})

describe('duplicate', () => {
  describe('disabled', () => {
    it('returns NOTHING_TO_DUPLICATE if there is no snapshot', () => {
      expect(
        duplicate.disabled({
          typeName: 'example',
          snapshots: {},
        } as unknown as OperationArgs),
      ).toBe('NOTHING_TO_DUPLICATE')
    })

    it("otherwise the operation isn't disabled", () => {
      expect(
        duplicate.disabled({
          typeName: 'example',
          snapshots: {
            draft: {} as SanityDocument,
            published: {} as SanityDocument,
          },
        } as unknown as OperationArgs),
      ).toBe(false)
    })
  })

  describe('execute', () => {
    it('creates a new document based on the source document', () => {
      const client = createMockSanityClient()

      duplicate.execute(
        {
          client,
          idPair: {
            draftId: 'drafts.my-id',
            publishedId: 'my-id',
          },
          snapshots: {
            draft: {
              _createdAt: '2021-09-14T22:48:02.303Z',
              _rev: 'exampleRev',
              _id: 'drafts.my-id',
              _type: 'example',
              _updatedAt: '2021-09-14T22:48:02.303Z',
              newValue: 'hey',
            },
          },
        } as unknown as OperationArgs,
        'my-duplicate-id',
      )

      expect(client.$log).toMatchSnapshot()
    })

    it('omits timestamp fields from the created document', () => {
      const client = createMockSanityClient()

      duplicate.execute(
        {
          client,
          idPair: {
            draftId: 'drafts.my-id',
            publishedId: 'my-id',
          },
          snapshots: {
            draft: {
              _createdAt: '2021-09-14T22:48:02.303Z',
              _rev: 'exampleRev',
              _id: 'drafts.my-id',
              _type: 'example',
              _updatedAt: '2021-09-14T22:48:02.303Z',
              newValue: 'hey',
            },
          },
        } as unknown as OperationArgs,
        'my-duplicate-id',
      )

      const creation = client.$log.observable.create.find(
        ([document]) => document._id === 'drafts.my-duplicate-id',
      )

      expect(creation[0]).not.toHaveProperty('_createdAt')
      expect(creation[0]).not.toHaveProperty('_updatedAt')
    })
  })

  it('applies the `mapDocument` function to the created document', () => {
    const client = createMockSanityClient()

    duplicate.execute(
      {
        client,
        idPair: {
          draftId: 'drafts.my-id',
          publishedId: 'my-id',
        },
        snapshots: {
          draft: {
            _createdAt: '2021-09-14T22:48:02.303Z',
            _rev: 'exampleRev',
            _id: 'drafts.my-id',
            _type: 'example',
            _updatedAt: '2021-09-14T22:48:02.303Z',
            newValue: 'hey',
          },
        },
      } as unknown as OperationArgs,
      'my-duplicate-id',
      {
        mapDocument: (document) => ({
          ...document,
          appendValue: 'appended',
        }),
      },
    )

    const creation = client.$log.observable.create.find(
      ([document]) => document._id === 'drafts.my-duplicate-id',
    )

    expect(client.$log).toMatchSnapshot()
    expect(creation[0].appendValue).toBe('appended')
  })
})
