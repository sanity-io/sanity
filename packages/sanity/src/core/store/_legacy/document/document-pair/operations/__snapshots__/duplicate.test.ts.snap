// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`duplicate > applies the \`mapDocument\` function to the created document 1`] = `
{
  "listen": [],
  "observable": {
    "action": [],
    "create": [
      [
        {
          "_id": "drafts.my-duplicate-id",
          "_rev": "exampleRev",
          "_type": "example",
          "appendValue": "appended",
          "newValue": "hey",
        },
        {
          "tag": "document.duplicate",
        },
      ],
    ],
    "fetch": [],
    "getDocuments": [],
    "listen": [],
    "request": [],
  },
  "request": [],
  "transaction": [],
}
`;

exports[`duplicate > execute > creates a new document based on the source document 1`] = `
{
  "listen": [],
  "observable": {
    "action": [],
    "create": [
      [
        {
          "_id": "drafts.my-duplicate-id",
          "_rev": "exampleRev",
          "_type": "example",
          "newValue": "hey",
        },
        {
          "tag": "document.duplicate",
        },
      ],
    ],
    "fetch": [],
    "getDocuments": [],
    "listen": [],
    "request": [],
  },
  "request": [],
  "transaction": [],
}
`;
